### PR TITLE
Allow cubes as input for `bias` preprocessor

### LINKED ISSUE #2082: Make `bias` preprocessor work with cubes
In it's current form, the [`bias`](https://github.com/ESMValGroup/ESMValCore/blob/291ce38cd6e77176b232de2a443044cb37f3e3fd/esmvalcore/preprocessor/_bias.py#L12) preprocessor can only be used on `PreprocessorFile` instances, which makes it hard to use outside of a recipe. It would be great to make this preprocessor work with cubes, too (e.g., like [`mask_multimodel`](https://github.com/ESMValGroup/ESMValCore/blob/bc5a4ac728c920f6d6c3e3676f8bc7b70a2bee5d/esmvalcore/preprocessor/_mask.py#L525)).

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR allows using cubes as input for the bias preprocessor, which drastically increases its usability outside of recipes. Additionally, I added a check if the `bias_type` argument is valid that is executed much earlier before processing datasets.

Closes #2082 

Link to documentation: https://esmvaltool--2183.org.readthedocs.build/projects/ESMValCore/en/2183/recipe/preprocessor.html#bias

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by valeriupredoi** [comment]
I'll have a review done tomorrow, bud

File: doc/recipe/preprocessor.rst [hunk 1]
```diff
@@ -2396,34 +2396,37 @@ true`` in the recipe, e.g.,
...
+  10^{-6}` kg m :math:`^{-2}` s :math:`^{-1}`) it is absolutely essential to
+  change the default value in order to get reasonable results.
+* ``keep_reference_dataset`` (:obj:`bool`, default: ``False``): If
+  ``True``, keep the reference dataset in the output. If ``False``, drop the
+  reference dataset.
+* ``exclude`` (:obj:`list` of :obj:`str`): Exclude specific datasets from
+  this preprocessor. Note that this option is only available in the recipe,
+  not when using :func:`esmvalcore.preprocessor.bias` directly (e.g., in
+  another python script). If the reference dataset has been excluded, an
+  error is raised.
```
Comment (valeriupredoi) [inline doc/recipe/preprocessor.rst:2429]: eh - that can be extended for the standalone use too, by looking up the dataset name in the cube metadata (not here, in a separate PR if you think that's a good idea)

Comment (valeriupredoi) [reply doc/recipe/preprocessor.rst:2429]: nice docs BTW :beer:

File: esmvalcore/preprocessor/_bias.py [hunk 2]
```diff
@@ -40,79 +66,123 @@ def bias(products, bias_type='absolute', denominator_mask_threshold=1e-3,
...
-    # the preprocessing chain of ESMValTool. To make sure that this
-    # preprocessor can also be used outside the ESMValTool preprocessing chain,
-    # an additional concatenate call is added here.
-    ref_cube = concatenate(reference_product.cubes)
+    # Get reference cube if not explicitly given
+    if ref_cube is None:
+        if all_cubes_given:
+            raise ValueError(
+                "`ref_cube` cannot be `None` when `products` is an iterable "
+                "of Cubes"
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_bias.py:98]: this is rather cryptic, could you please expand a bit for the user's sake?

File: esmvalcore/preprocessor/_bias.py [hunk 3]
```diff
@@ -40,79 +66,123 @@ def bias(products, bias_type='absolute', denominator_mask_threshold=1e-3,
...
+
+    return (ref_cube, ref_product)
+
+
+def _calculate_bias(cube: Cube, ref_cube: Cube, bias_type: BiasType) -> Cube:
+    """Calculate bias for a single cube relative to a reference cube."""
+    cube_metadata = cube.metadata
+
+    if bias_type == 'absolute':
+        cube = cube - ref_cube
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_bias.py:174]: does iris now allow cube subtractions when cubes have differing metadata?

File: tests/integration/recipe/test_recipe.py [hunk 4]
```diff
@@ -2767,3 +2767,110 @@ def test_invalid_stat_preproc(tmp_path, patched_datafinder, session):
...
+        """)
+    msg = (
+        "Expected exactly 1 dataset with 'reference_for_bias: true' in "
+        "products"
+    )
+    with pytest.raises(RecipeError) as exc:
+        get_recipe(tmp_path, content, session)
+    assert str(exc.value) == INITIALIZATION_ERROR_MSG
+    assert msg in exc.value.failed_tasks[0].message
+    assert "found 2" in exc.value.failed_tasks[0].message
```
Comment (valeriupredoi) [inline tests/integration/recipe/test_recipe.py:2841]: slightly confused as to how this test differs from the one above, given there are identical settings for `bias` preproc

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
looking very good, and cheers, Manu! The two reasons why I'm not approving straightaway are that error that I think needs be a bit more explicit, and (for the life of me) I can't see how those two integration tests differ :beer:

File: doc/recipe/preprocessor.rst [hunk 5]
```diff
@@ -2396,34 +2396,37 @@ true`` in the recipe, e.g.,
...
+  calculate bias as :math:`\frac{X - R}{R}`).
+* ``denominator_mask_threshold`` (:obj:`float`, default: ``1e-3``):
+  Threshold to mask values close to zero in the denominator (i.e., the
+  reference dataset) during the calculation of relative biases. All values
+  in the reference dataset with absolute value less than the given threshold
+  are masked out. This setting is ignored when ``bias_type`` is set to
+  ``'absolute'``. Please note that for some variables with very small
+  absolute values (e.g., carbon cycle fluxes, which are usually :math:`<
+  10^{-6}` kg m :math:`^{-2}` s :math:`^{-1}`) it is absolutely essential to
+  change the default value in order to get reasonable results.
```
[the remark this answers is not in the document]
Comment (schlunma) [reply doc/recipe/preprocessor.rst:2423]: If the value is very low, your relative bias might end up with huge numbers, and you won't see anything in a plot apart from that big number. If your value is too high, you mask out too much data. It's always a tradeoff.

File: doc/recipe/preprocessor.rst (hunk 1 above)
Comment (schlunma) [reply doc/recipe/preprocessor.rst:2431]: True, but in the standalone use I would argue that people can just pass only the relevant cubes to this preprocessor.

Comment (valeriupredoi) [reply doc/recipe/preprocessor.rst:2431]: that too!

File: tests/integration/recipe/test_recipe.py (hunk 4 above)
Comment (schlunma) [reply tests/integration/recipe/test_recipe.py:2841]: The one at the top uses

```yaml
additional_datasets:
  - {dataset: CanESM5}
  - {dataset: CESM2}
```

This one here:

```yaml
additional_datasets:
  - {dataset: CanESM5, reference_for_bias: true}
  - {dataset: CESM2, reference_for_bias: true}
```

File: esmvalcore/preprocessor/_bias.py (hunk 2 above)
Comment (schlunma) [reply esmvalcore/preprocessor/_bias.py]: I tried to improve that in https://github.com/ESMValGroup/ESMValCore/pull/2183/commits/3b8f7d33564bf1965191d74d76512985582f9138

File: doc/recipe/preprocessor.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (valeriupredoi) [reply doc/recipe/preprocessor.rst:2423]: hmm yes - good points - but I'd rather we output some completely mad plots that will make the user figure out that they did something wrong, than have "decent" but wrong plots and numbers - my take is default should be something outrageous in these cases, so we force the user to change it. Your call ->

File: esmvalcore/preprocessor/_bias.py (hunk 3 above)
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_bias.py:174]: no, that what you mentioned. There was a time in the past when iris was disallowing cube subtractions even if it rained outside

File: doc/recipe/preprocessor.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (schlunma) [reply doc/recipe/preprocessor.rst:2425]: I don't think there is a right or wrong here, it's either very large numbers or a too large mask - both are not optimal. I think something in the middle is reasonable. Either way (to high or too low) will look outrageous, so the user will definitely want to change it. I think we should keep it.

File: tests/integration/recipe/test_recipe.py (hunk 4 above)
Comment (valeriupredoi) [reply tests/integration/recipe/test_recipe.py:2841]: ah there we go, my blindness knows no boundaries, thanks, bud

File: doc/recipe/preprocessor.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (valeriupredoi) [reply doc/recipe/preprocessor.rst:2425]: I think it's OK, downside of having me as a reviewer - I tangent away on philosophical matters :grin:

**Review by valeriupredoi — APPROVED** [review]
many thanks for addressing the review points, bud :beer:

File: esmvalcore/preprocessor/_bias.py (hunk 2 above)
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_bias.py]: biutiful!

**Comment by schlunma** [comment]
Thanks for the review V! Merging this now since I need it to continue with https://github.com/ESMValGroup/ESMValCore/pull/2299.

**Pull request merged**